File: Python/Archive/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(next=head)
        prev, back, front = dummy, head, head
        for _ in range(n - 1): front = front.next
        while front.next:
            prev = back
            front = front.next
            back = back.next
        prev.next = back.next
        return dummy.next

File: C++/Two Pointers/19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy, *front = head, *back = head;
        for (int i = 0; i < n-1; i++) front = front->next;
        while (front->next) {
            prev = back;
            back = back->next;
            front = front->next;
        }
        prev->next = back->next;
        return dummy->next;
    }
};